### PR TITLE
rgw: add rgw_compression_type=random for teuthology testing

### PR DESCRIPTION
**Pull request by cbodley**
adds support for compression type "random" to the `Compressor::create()` factory, similar to the strategy for ms_type=random in the `Messenger::create()` factory

rgw uses this, along with some extra logic to make sure that:
* all buffers in an upload are compressed with the same plugin, and
* all parts of a multipart upload use the same plugin

this will be used in the rgw/verify suite of ceph-qa-suite (see https://github.com/ceph/ceph-qa-suite/pull/977)

### CODE REVIEW

**Comment by cbodley** [comment]
scheduled at http://pulpito.ceph.com/cbodley-2016-11-10_14:57:29-rgw:verify-wip-rgw-compression-type-random---basic-mira

**Comment by cbodley** [comment]
@aclamk could you please help review?

(@ifed01 may also be interested)

File: src/compressor/Compressor.cc [hunk 1]
```diff
@@ -59,6 +59,21 @@ boost::optional<Compressor::CompressionMode> Compressor::get_comp_mode_type(cons
...
+    static std::random_device seed;
+    static std::default_random_engine engine(seed());
+    static Spinlock mutex;
+
+    int alg = COMP_ALG_NONE;
+    std::uniform_int_distribution<> dist(0, COMP_ALG_LAST - 1);
+    {
+      std::lock_guard<Spinlock> lock(mutex);
+      alg = dist(engine);
+    }
```
Comment (cbodley) [inline src/compressor/Compressor.cc]: seeing a lot of spam like this: `/usr/lib64/ceph/libceph_none.so: cannot open shared object file: No such file or directory`

if we get `alg == COMP_ALG_NONE` here, it should probably return nullptr directly

**Comment by aclamk** [comment]
Looks good.

**Comment by cbodley** [comment]
clean teuthology run at http://pulpito.ceph.com/cbodley-2016-11-14_11:43:52-rgw:verify-wip-rgw-compression-type-random---basic-mira/

(only failures are valgrind and HTTPConnectionPool timeouts)

**Comment by cbodley** [comment]
jenkins test this please

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -2808,6 +2808,29 @@ int RGWPutObj::get_data(const off_t fst, const off_t lst, bufferlist& bl)
   return ret;
 }
 
+// special handling for rgw_compression_type = "random" with multipart uploads
+static CompressorRef get_compressor_plugin(const req_state *s)
+{
+  const auto& compression_type = s->cct->_conf->rgw_compression_type;
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc]: That certainly sounds like a great idea, but are there any other policy options we should be considering?

**Pull request merged**